README.md
# The Concordant Service

[![License](https://img.shields.io/badge/license-MIT-green)](https://opensource.org/licenses/MIT)
[![pipeline status](https://gitlab.inria.fr/concordant/software/c-service/badges/master/pipeline.svg)](https://gitlab.inria.fr/concordant/software/c-service/-/commits/master)
[![coverage report](https://gitlab.inria.fr/concordant/software/c-service/badges/master/coverage.svg)](https://gitlab.inria.fr/concordant/software/c-service/-/commits/master)

The C-Service is a lightweight proxy service that can run on many kinds of devices and connect in one end C-Client instances running on edge devices (terminal or PoP) to support transactions and replication between services, and on the other end any backend server that is adapted to its drivers interface to store and replicate data.

Currently on beta version, it runs as a ServiceWorker in web browsers,
using PouchDB to synchronize with a remote CouchDB server.
It can also be launched as a standalone application, typically server-side,
to be used by browsers that do not support Service Workers.

The C-Service exposes a REST API which should not be accessed directly,
as it may not be stable; use the [c-client library](https://gitlab.inria.fr/concordant/software/c-client/) instead.

*Note on current implementation:* the C-Service beta version is a distributed database service for CRDTs of the [C-CRDTlib](https://github.com/concordant/c-crdtlib).

## Getting started

### Requirements

For the next steps, you will need the following software:

- Make sure you have the latest stable version of [Node.js](https://nodejs.org/en/download/);
- C-Service synchronizes with a [CouchDB](https://docs.couchdb.org/en/stable/install/index.html) backend.
- Once CouchDB installed, use the [CouchDB user panel (Fauxton)](http://127.0.0.1:5984/_utils/) to setup a database and a user.

In addition, you will need to enable CORS in CouchDB, using Fauxton
or the [dedicated NPM package from PouchDB](https://github.com/pouchdb/add-cors-to-couchdb):

```shell
npx add-cors-to-couchdb [http://mycouchDBserver.org] \
    [-u myusername] [-p mypassword]
(server defaults to localhost)
```

### Use the C-Service as a Service Worker

A Service Worker implementation is bundled with the C-Service package, in `dist/c-service-worker.js`.

To use it, just copy (or link) it to an appropriate location
and register it as a ServiceWorker in your application.
For more information, see our example applications.

This Service Worker is configured to use our demo database,
so feel free to re-build it to use your own:
adjust the configuration in `src/config.json` and run

```shell
npm install
```

In a future release, the C-Client will be able to provide these settings at runtime.

### Start C-Service standalone

Ensure CouchDB is running before starting C-Service.
If it is running on another host, set the `COUCHDB_URL` accordingly.
You will also need to provide CouchDB credentials:

```bash
export COUCHDB_URL=my-couchdb-server.org
export COUCHDB_USER=my-user COUCHDB_PASSWORD=my-passwd
```

It is also possible to configure WebSocket path used by the service (default is `/`):

```bash
export WS_PATH=my-websocket-path
```

Then run the service:

```bash
npx @concordant/c-service
```

This launches an Express server listening on TCP port 4000 and a WebSocket server on port 8999.

## Setup C-Service as a Systemd service

First ensure CouchDB is setup as a systemd service named `couchdb.service`
(or adapt the name in `c-service.service` file).

Put the c-service folder in `/opt/` or adapt the `c-service.service` file.

Put the provided file `c-service.service` in `/etc/systemd/system/`.

Create the file `/etc/systemd/system/c-service.conf`
containing CouchDB user credentials:

```shell
COUCHDB_USER=<couchdb_id>
COUCHDB_PASSWORD=<couchdb_password>
```

Then start the service:

```shell
systemctl start c-service
```

## Test C-Service using curl

The standalone C-Service exposes a REST API on `/api`,
whose description can be found in file `swagger.yml`.

Once your C-Service is running you can query its REST API with curl.
You can view your changes in the [CouchDB user pannel](http://127.0.0.1:5984/_utils/).

Create the `myapp` database:

```bash
curl --request POST                               \
     --header "Content-Type: application/json"    \
     --data '{"appName":"myapp"}'                 \
     http://127.0.0.1:4000/api/create-app
```

Insert/update a PNCounter named `myobject` in `myapp` database:

```bash
curl --request POST                               \
     --header "Content-Type: application/json"    \
     --data '{"appName":"myapp","id":"{\"name\":\"myobject\",\"type\":\"PNCounter\"}","document":"{\"type\":\"PNCounter\",\"metadata\":{\"increment\":[{\"name\":\"clientid\"},{\"first\":60,\"second\":{\"uid\":{\"name\":\"clientid\"},\"cnt\":-21474836}}],\"decrement\":[]},\"value\":60}"}'  \
     http://127.0.0.1:4000/api/update-object
```

Get the PNCounter named `myobject` in `myapp` database:

```bash
curl --request POST                               \
     --header "Content-Type: application/json"    \
     --data '{"appName":"myapp","id":"{\"name\":\"myobject\",\"type\":\"PNCounter\"}"}'                       \
     http://127.0.0.1:4000/api/get-object
```

Get all objects of `myapp` database:

```bash
curl --request POST                               \
     --header "Content-Type: application/json"    \
     --data '{"appName":"myapp"}'                 \
     http://127.0.0.1:4000/api/get-objects
```

Delete `myapp` database:

```bash
curl --request POST                               \
     --header "Content-Type: application/json"    \
     --data '{"appName":"myapp"}'                 \
     http://127.0.0.1:4000/api/delete-app
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Added
- Changed
- Deprecated
- Removed
- Fixed
- Security

## [1.2.4] - 2022-09-10

- Fixed
  - Build deps error

## [1.2.3] - 2022-09-10

- Changed
  - Update License year
  - Upgrade NPM dependencies
  - Moved to Github: update links and Actions

## [1.2.2] - 2021-12-02

- Added
  - WebSocket path is now configurable

## [1.2.1] - 2021-08-04

- Added
  - Detect and resolve eventual conflicts
  - Ensure that StoreDataSource exists in the service worker

## [1.2.0] - 2021-07-09

- Added
  - Add subscription and web socket in DataSource
  - Add StoreDataSource class to store DataSource
  - Add subscription endpoint
  - Run websocket server on port 8999 (default port)
  - Add subscription support in service worker

## [1.1.7] - 2021-04-16

- Changed
  - updateObject confirms the update
  - Change references from Github to Gitlab

## [1.1.6] - 2021-04-01

- Changed
  - Put GraphQL schema in a separate file
  - Systemd service is using npx with package from npmjs
  - Clean and update the code abstraction over PouchDB
  - Use our PouchDB datasource
- Removed
  - Remove unused packages
- Fixed
  - Fix tests to match the new typing interface


src/service.ts
#!/usr/bin/env node

/**
 * MIT License
 *
 * Copyright (c) 2022, Concordant and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { ApolloServer, gql } from "apollo-server-express";
import * as bodyParser from "body-parser";
import cors from "cors";
import express from "express";
import * as http from "http";
import * as WebSocket from "ws";
import { makeExecutableSchema } from "graphql-tools";
import { OpenAPI, useSofa } from "sofa-api";
import { graphQLSchema } from "./schema";
import { PouchDBParams } from "./database/PouchDBDataSource";
import StoreDataSource from "./database/StoreDataSource";

const db = new StoreDataSource();

// CouchDB variables
const url = process.env.COUCHDB_URL || "http://localhost:5984/";
const username = process.env.COUCHDB_USER || undefined;
const password = process.env.COUCHDB_PASSWORD || undefined;

// GraphQL definition
const typeDefs = gql(graphQLSchema);

// Resolvers for GraphQL
const resolvers = {
  Mutation: {
    createApp(_: any, { appName }: any): Promise<boolean> {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.isConnected();
    },
    deleteApp(_: any, { appName }: any): Promise<boolean> {
      return Promise.reject("Not yet implemented");
    },
    updateObject: (_: any, { appName, id, document }: any): Promise<string> => {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.updateObject(id, document);
    },
    subscribe: (
      _: any,
      { appName, collectionUId, userId }: any
    ): Promise<boolean> => {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.subscribe(collectionUId, userId);
    },
    unsubscribe: (
      _: any,
      { appName, collectionUId, userId }: any
    ): Promise<boolean> => {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.unsubscribe(collectionUId, userId);
    },
    replicator: (_: any, { source, target, continuous }: any) => {
      return Promise.reject("Not yet implemented");
    },
  },
  Query: {
    getObjects: (_: any, { appName }: any): Promise<Promise<string>[]> => {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.getObjects();
    },
    getObject: (_: any, { appName, id }: any): Promise<string> => {
      const params: PouchDBParams = {
        url,
        username,
        password,
        dbName: appName,
      };
      const dataSource = db.getOrConnect(params);
      return dataSource.getObject(id);
    },
    replicators: async () => {
      return Promise.reject("Not yet implemented");
    },
    replicator: async (_: any, { id }: any) => {
      return Promise.reject("Not yet implemented");
    },
  },
};

// Create express server
const app = express();

// Cors configurations
const coreOptions: cors.CorsOptions = {
  allowedHeaders: [
    "Origin",
    "X-Requested-With",
    "Content-Type",
    "Accept",
    "X-Access-Token",
  ],
  credentials: true,
  methods: "GET,HEAD,OPTIONS,PUT,PATCH,POST,DELETE",
  origin: "*",
  preflightContinue: false,
};

// Configure API
const schema = makeExecutableSchema({
  resolvers,
  typeDefs,
});
const openApi = OpenAPI({
  info: {
    title: "Concordant API",
    version: "1.0.0",
  },
  schema,
});
openApi.save("./swagger.yml");

// Bind express server and API
app.use(
  "/api",
  cors(coreOptions),
  bodyParser.json({ limit: "10mb" }),
  useSofa({
    schema,
    onRoute(info) {
      openApi.addRoute(info, {
        basePath: "/api",
      });
    },
    method: {
      "Query.getObjects": "POST",
      "Query.getObject": "POST",
    },
  })
);

// Configure GraphQL server
const serverApollo = new ApolloServer({
  resolvers,
  typeDefs,
});

// Bind GraphQL server with API
serverApollo.applyMiddleware({ app });

// Launch server
app.listen(4000, "0.0.0.0");

//initialize a simple http server
const server = http.createServer(app);

//initialize the WebSocket server instance
const wss = new WebSocket.Server({
  server: server,
  path: process.env.WS_PATH || "/",
});

wss.on("connection", (ws: WebSocket) => {
  ws.on("message", (message: string) => {
    const obj = JSON.parse(message);
    const params: PouchDBParams = {
      url,
      username,
      password,
      dbName: obj.appName,
    };
    const dataSource = db.getOrConnect(params);
    dataSource.addWebSocket(obj.userId, ws);
  });
  ws.on("close", (code: number, reason: string) => {
    // do nothing
  });
});

//start our web socket server
server.listen(process.env.PORT || 8999);
